### PR TITLE
DBZ-402 Reader now has ability to initialize resource prior to starting

### PR DESCRIPTION
**Pull request by rhauch**
See #309 for the original suggested fix for DBZ-402. 

This alternative, IMO, is a bit cleaner because the `Reader` contract implies that implementations that are not started need not be stopped. Both the `BinlogReader` and `SnapshotReader` do not satisfy this, since they both register their JMX metrics upon construction and not upon start. The problem is that if they were to do it upon start, when a connector is performing a snapshot and then reading the binlog, the `BinlogReader` is not started until after the `SnapshotReader` is finished, which means the JMX metrics would not be registered until after the snapshot completed. This is counterintuitive.

So, I think the `Reader` interface is missing a method that signals that the reader implementation can initialize resources but prior to `start()`. This PR adds an `initialize()` method to represent this condition, and `initialize()` is to be called before `start()`. For example, implementations that are combinations of other readers (e.g., `ChainedReader`) can implement `initialize()` by immediately calling `initialize()` on all of its readers, whereas the readers are still started and stopped sequentially as they were before.

This makes it easier to correctly register JMX metrics, for example, up front once all of the readers have been configured for use, rather than to do so upon construction when it is not clear that a particular Reader will be used after it is constructed.

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java [hunk 1]
```diff
@@ -69,7 +70,15 @@ public void uponCompletion(Runnable handler) {
...
     @Override
+    public void initialize() {
+        if (this.initialized.compareAndSet(false, true)) {
+            doInitialize();
+        }
+    }
+
+    @Override
     public void start() {
+        initialize();
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java]: It's surprising that it is called here by `start()`. I'd rather have expected that the task or connector calls it before calling `start()` (as you actually do it below)?

Comment (rhauch) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java]: It needs to be called, and I debated about not calling it here. But all current implementation handle calling more than once, so I thought this was safer. But I don’t have a preference.

**Comment by rhauch** [comment]
The only reason I referred to it as a PoC is that I wanted feedback on the direction as this seemed cleaner to me. But I don’t feel strongly about it.

**Comment by pgoranss** [comment]
I agree with this approach. The BinlogReader JMX metrics will now only register if that reader is added. I think it will resolve the issue, I'll test it out today.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java]: Do you mean to say "It needs *not* to be called"? If so, I'd prefer to not do it.

**Comment by rhauch** [comment]
@pgoranss, thanks for the feedback. And just to be clear, the JMX metrics would only be registered once the reader is _*started*_.

@gunnarmorling we just need to resolve your question about whether `initialize()` should be called automatically by `start()` if not called explicitly beforehand. This is an internal API, so we can keep things simple by requiring it be called before start. Let me know what you think.

**Comment by gunnarmorling** [comment]
> The only reason I referred to it as a PoC is that I wanted feedback 

Ok, just wasn't sure anything is missing :) It looks good and is the route we should be taking IMO.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java (hunk 1 above)
Comment (rhauch) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/AbstractReader.java]: Sorry, meant to say that `initialize()` needs to be called at some point, either explicitly by the user or, if we wanted, implicitly in the `start()` method.

**Comment by gunnarmorling** [comment]
> This is an internal API, so we can keep things simple by requiring it be called before start

+1

**Comment by gunnarmorling** [comment]
Hum, so taking a step, I'm wondering why we not simply change it so to only instantiate the `BinlogReader` if it's actually going to be added to the chained readers?

That seems the much simpler fix, but then the `instantiate()` contract could be useful either way, so both works for me.

**Comment by pgoranss** [comment]
@gunnarmorling That was another more direct approach I was considering as well. 

I think I like the extensibility of the instantiate() contract and I have tested this locally and it fixes the issue I saw in DBZ-402 so I'm good with it.

**Comment by rhauch** [comment]
I agree that only instantiating it if/when needed is the easier approach, but that becomes a bit harder and uglier in the code. Personally, I think it's easier to build up the instances as desired and to initialize and use them only if needed. I also think that it may make setting up the readers a bit easier when we get to DBZ-175.

I'll update the PR to document that `initialize()` must be called prior to `start()`, and is expected to only be called once for each reader.

**Comment by rhauch** [comment]
@gunnarmorling @pgoranss Updated the PR. Ready to be merged if approved.

**Comment by gunnarmorling** [comment]
@rhauch looks great, thanks! The only thing remaining is that `AbstractReader#start()` still calls `initialize()` which it shouldn't as per the discussion above. I'll remove that invocation and merge.

**Comment by gunnarmorling** [comment]
Rebased, applied and added one more commit to remove the second call to `initialize()`. Thanks, @rhauch!

**Pull request closed without merging**